### DISCUSSION TITLE
15MB .tflite files with @react-native-firebase/ml

### DISCUSSION THREAD

**Discussion by MujtabaFR** [invertase/react-native-firebase#4750] [Q&A] [answered]
Hello,
I'm trying to use **Text Recognition** (in the cloud) in my app, So I had to install `@react-native-firebase/ml` .. but I was surprised that the final apk size **increased by 15MB** from `@react-native-firebase/ml` alone.

I know that `@react-native-firebase/ml` is specialized to "**Cloud**" inferences, while `react-native-mlkit` is specialized to "**On-Device**" inferences. So Why all these `.tflite` files attached to the `@react-native-firebase/ml` library ?

Is it a bug ? is there anyway I can use cloud Text Recognition without these 15MB `.tflite` files ?

Thanks

![stack](https://user-images.githubusercontent.com/7612635/103988826-c7488280-519f-11eb-8dfd-b9f078b5da25.jpg)

**Comment by mikehardy** [comment]
It is likely a bug as the on-device ML stuff was removed from react-native-firebase (by me, so likely my oversight)

Something about these:

https://github.com/invertase/react-native-firebase/blob/83a7382068e8d5235d2aa970bd7d6b205d64728b/packages/ml/android/build.gradle#L95-L101

Not sure which one is bringing down the local models but you could experiment by bisecting a removal of them and making a PR of the result after discovering which ones are no longer necessary?

**Comment by MujtabaFR** [comment, marked as answer]
The `.tflite` files were removed when removing these two lines

https://github.com/invertase/react-native-firebase/blob/83a7382068e8d5235d2aa970bd7d6b205d64728b/packages/ml/android/build.gradle#L97

https://github.com/invertase/react-native-firebase/blob/83a7382068e8d5235d2aa970bd7d6b205d64728b/packages/ml/android/build.gradle#L100

After removing both lines, apk size was reduces by **10MB**

I created a [PR](https://github.com/invertase/react-native-firebase/pull/4753) removing them, as face detection is part of [ML Kit](https://firebase.google.com/docs/ml#ml_kit_ready-to-use_on-device_models) as on-device model

**Reply by mikehardy** [reply]
Fantastic @MujtabaFR - your PR is already merged and as soon as CI finishes chewing on it I'll hit the button to publish 10.4.1 with your fix 🏆  😄

**Comment by MujtabaFR** [comment]
**face detection** had `.tflite` models of **10mb** size

**image labeling** has **4mb** `.so` libraries, loaded by these two lines :

https://github.com/invertase/react-native-firebase/blob/83a7382068e8d5235d2aa970bd7d6b205d64728b/packages/ml/android/build.gradle#L98-L99

![libs](https://user-images.githubusercontent.com/7612635/104085396-27066280-5260-11eb-891d-9d0c8321ac8f.jpg)

I'm not sure if these two lines are necessary for `@react-native-firebase/ml` or not

But I'm removing them from my app. And mentioning it here so you have an idea about them @mikehardy 👌

**Reply by mikehardy** [reply]
I believe those are necessary to handle downloaded ML models and are exactly the point of the Firebase ML component that is still here (and not moved to MLKit) ?

**Reply by MujtabaFR** [reply]
According to the documentation.. If you want to use **image labeling**, there are two options :
- **Cloud** image labeling (Firebase ML)
   To use it you need to add this dependencies : ([docs](https://firebase.google.com/docs/ml/android/label-images#before-you-begin))
  ```
  implementation platform('com.google.firebase:firebase-bom:26.2.0')
  implementation 'com.google.firebase:firebase-ml-vision'
  ```

- **On-device** image labeling (ML Kit)
   You have two options to use it : ([docs](https://developers.google.com/ml-kit/vision/image-labeling/android#before-you-begin))
   - Either bundling the model in your app : 
`implementation 'com.google.mlkit:image-labeling:17.0.1'`
   - Or using model within Google Play Service (configure your app to automatically download the model to the device after your app is installed from the Play Store)
`implementation 'com.google.android.gms:play-services-mlkit-image-labeling:16.0.1'`
   
   

### In summery:
`@react-native-firebase/ml` is specialized in **Cloud** inferences, so all the processes **will be done in the cloud**.
(it doesn't need to have any model bundled with the app, or to have any library to handle downloaded models).

So you can remove these two lines and save **4mb** for the users 🙌  😁
https://github.com/invertase/react-native-firebase/blob/83a7382068e8d5235d2aa970bd7d6b205d64728b/packages/ml/android/build.gradle#L98-L99

**Reply by mikehardy** [reply]
I think you are right about that. In fact we could probably (at least temporarily) get rid of the firebase-ml-model-interpreter one as well as we *want* to do custom inference here on custom models - that is a firebase API surface area (https://firebase.google.com/docs/ml/android/use-custom-models) but we haven't implemented it yet, so for cloud-ML purposes it is dead weight I think

**Reply by mikehardy** [reply]
v10.6.4 just shipped with this change, 4MB people don't need, deleted...
